### PR TITLE
[Feature] Implemented a Mute Reason referenced in Issue #385

### PR DESCRIPTION
**Pull request by LCookman**
We have implemented the ability to add a reason when muting a player. The below list of commands are valid command executions, note that this works just like the previous mute, except with a reason. Some drawbacks of the current implementation are below the commands.
```
- /mute <player> [dateDiff] [reason]
- /mute <player> [dateDiff]
- /mute <player> [reason]
- /mute <player>
```
This introduced some new cases to handle. In this version muting with a reason, but no time does not toggle mute like muting without a time, before a reason was an option `[see example 1]`. This is easy to change if you feel that if a player is muted, they should be unmuted if example 1 is run while that player is muted. Also, having an invalid time no longer throws an exception or cancels the command. It simply sets the mute reason to what comes after the name `[see example 2]`. If the user misses a space after the time and includes part of the reason with the time, that portion of the reason will not make it to the actual reason `[see example 3]`, Which will read `Player nightwarrier muted for 9 minutes 59 seconds. Reason: im rico`
```
Examples:
1. /mute delbertina This is a mute
2. /mute CreedTheFreak *d Whoops
3. /mute nightwarrier 10mheyo im rico
```
If you would like to test out this implementation further we have included a compiled Jar file within the zip file. If you have any questions we would be happy to answer them.
[EssentialsX-2.0.1.zip](https://github.com/drtshock/Essentials/files/1309457/EssentialsX-2.0.1.zip)

### CODE REVIEW

**Comment by delbertina** [comment]
I was curious if this isn't an acceptable way to implement this feature. We could also add a config value or permission to check for. Something along the lines of "allow-mute-reason" as a config value or "essentials.mute.reason" as a permission to use mute reasons.

File: Essentials/src/com/earth2me/essentials/EssentialsPlayerListener.java [hunk 1]
```diff
@@ -80,7 +80,14 @@ public void onPlayerChat(final AsyncPlayerChatEvent event) {
         final User user = ess.getUser(event.getPlayer());
         if (user.isMuted()) {
             event.setCancelled(true);
-            user.sendMessage(tl("voiceSilenced"));
+
+            if (user.getMuteReason ().equals ("")) {
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/EssentialsPlayerListener.java]: The space between `getMuteReason` and its brackets doesn't match the style conventions followed by the rest of EssentialsX's code.

File: Essentials/src/com/earth2me/essentials/UserData.java [hunk 2]
```diff
@@ -510,6 +512,30 @@ public void setMuted(boolean set) {
...
+    public String getMuteReason() {
+        if (muteReason != null) {
+            return muteReason;
+        }
+        else {
+            return "";
+        }
+    }
+
+    public void setMuteReason (String reason) {
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/UserData.java]: The space here is also inconsistent with the rest of EssX.

File: Essentials/src/com/earth2me/essentials/commands/Commandmute.java [hunk 3]
```diff
@@ -66,17 +79,41 @@ public void run(final Server server, final CommandSource sender, final String co
 
             if (muted) {
                 if (muteTimestamp > 0) {
-                    sender.sendMessage(tl("mutedPlayerFor", user.getDisplayName(), muteTime));
-                    user.sendMessage(tl("playerMutedFor", muteTime));
+                    if (user.getMuteReason ().equals ("")) {
+                        sender.sendMessage(tl("mutedPlayerFor", user.getDisplayName(), muteTime));
+                        user.sendMessage(tl("playerMutedFor", muteTime));
+                    }
+                    else {
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandmute.java]: Another style issue - `else`s should be on the same line as the previous block's closing bracket.

File: Essentials/src/com/earth2me/essentials/commands/Commandafk.java [hunk 4]
```diff
@@ -51,7 +51,12 @@ public void run(Server server, CommandSource sender, String commandLabel, String
     private void toggleAfk(User sender, User user, String message) throws Exception {
         if (message != null && sender != null) {
             if (sender.isMuted()) {
-                throw new Exception(tl("voiceSilenced"));
+                if (sender.getMuteReason ().equals ("")) {
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandafk.java]: I can't help but feel that there is a better way to format these than adding if/else statements everywhere and concatenating the string inside those blocks. Personally, I would:
* Add a `hasMuteReason` method to `UserData` which does `getMuteReason().equals("")`;
* Use a ternary operator when concatenating the strings instead of if/else blocks:
```java
throw new Exception(tl("voiceSilenced") + (sender.hasMuteReason() ?
                tl("muteReason", sender.getMuteReason())));
```

File: Essentials/src/messages.properties [hunk 5]
```diff
@@ -597,3 +598,4 @@ createKitFailed=\u00a74Error occurred whilst creating kit {0}.
 createKitSeparator=\u00a7m-----------------------
 createKitSuccess=\u00a76Created Kit: \u00a7f{0}\n\u00a76Delay: \u00a7f{1}\n\u00a76Link: \u00a7f{2}\n\u00a76Copy contents in the link above into your config.yml.
 whoisUuid=\u00a76 - UUID\:\u00a7r {0}
+muteFormat=\u00a76 Reason: \u00a7c{0}
```
Comment (mdcfe) [inline Essentials/src/messages.properties]: This is a duplicate `muteFormat` key.

File: Essentials/src/messages.properties [hunk 6]
```diff
@@ -494,6 +494,7 @@ vanished=\u00a76You are now completely invisible to normal users, and hidden fro
 versionMismatch=\u00a74Version mismatch\! Please update {0} to the same version.
 versionMismatchAll=\u00a74Version mismatch\! Please update all Essentials jars to the same version.
 voiceSilenced=\u00a76Your voice has been silenced\!
+muteFormat=\u00a74 Reason: {0}
```
Comment (mdcfe) [inline Essentials/src/messages.properties]: Having spaces before the string here means that the string can only be used when concatenating it with another string. In addition, calling it `muteFormat` doesn't seem very consistent with the other uses of `-Format` - `muteReason` would be a more descriptive name for the string.

**Review by mdcfe — CHANGES_REQUESTED** [review]

**Comment by delbertina** [comment]
We have implemented the changes requested, if we missed something please let us know. In taking out the space from the muteReason message, the places that used it required a space to be added. We fixed this, by amending a space before adding the mute reason in the places that needed it. We don’t intend for this change to be accepted, but we don’t know what way of fixing this would best fit the style of this project. The following are some idea we came up.

-Add a message “muteReasonAmend” that is “muteReason” but with a space before to be used when amending onto.
-Similar to how ban reasons are shown, have a “...Reason” version of all the messages that can have a reason in them in the messages.properties files.
-Keep the amended space like how we have it now.

**Comment by mdcfe** [comment]
I would suggest the second option, with strings like `voiceSilencedReason` for example, as it would be consistent with other strings in EssentialsX. It also makes formatting more flexible for server owners should they wish to alter it.

**Comment by LCookman** [comment]
Alrighties, we think that we have the needed bits and bobs implemented. If not please let us know and we shall make haste to add those in or fix the things. We implemented the second option of adding the various mute reasons within the .properties files.

**Review by LaxWasHere — APPROVED** [review]

File: Essentials/src/com/earth2me/essentials/User.java [hunk 7]
```diff
@@ -537,6 +537,7 @@ public boolean checkMuteTimeout(final long currentTime) {
                 setMuteTimeout(0);
                 sendMessage(tl("canTalkAgain"));
                 setMuted(false);
+                setMuteReason ("");
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/User.java]: Formatting - we don't use spaces between method names and brackets.

File: Essentials/src/com/earth2me/essentials/UserData.java [hunk 8]
```diff
@@ -73,6 +73,7 @@ public final void reloadConfig() {
         godmode = _getGodModeEnabled();
         muted = _getMuted();
         muteTimeout = _getMuteTimeout();
+        muteReason = _getMuteReason ();
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/UserData.java]: Formatting again.

File: Essentials/src/com/earth2me/essentials/UserData.java [hunk 9]
```diff
@@ -510,6 +512,34 @@ public void setMuted(boolean set) {
...
 
+    public String _getMuteReason() {
+        return config.getString("muteReason");
+    }
+
+    public String getMuteReason() {
+        if (muteReason != null) {
+            return muteReason;
+        }
+        else {
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/UserData.java]: Formatting - `else`s should be on the same line.

Comment (mdcfe) [reply Essentials/src/com/earth2me/essentials/UserData.java]: Also, why not just return `muteReason` even if it *is* null? We have `hasMuteReason` anyway if someone wants to check whether there is a mute reason in advance, and using `null` has an actual symbolic meaning making it clearer what the returned value means.

File: Essentials/src/com/earth2me/essentials/UserData.java [hunk 10]
```diff
@@ -510,6 +512,34 @@ public void setMuted(boolean set) {
...
+            muteReason = null;
+        } else {
+            muteReason = reason;
+            config.setProperty ("muteReason", reason);
+        }
+        config.save();
+    }
+
+    public boolean hasMuteReason(){
+        return !getMuteReason().equals("");
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/UserData.java]: Is there any particular reason to call `getMuteReason` instead of just checking the value of `muteReason` directly?

File: Essentials/src/com/earth2me/essentials/commands/Commandmute.java [hunk 11]
```diff
@@ -50,11 +50,24 @@ public void run(final Server server, final CommandSource sender, final String co
...
+                String muteReason;
+                try {
+                    muteTimestamp = DateUtil.parseDateDiff(time, true);
+                    muteReason = getFinalArg (args, 2);
+                } catch (Exception e) {
+                    user.setMuted ((!user.getMuted ()));
+                    muteReason = getFinalArg (args, 1);
+                }
+
+                user.setMuteReason (muteReason);
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandmute.java]: Format.

File: Essentials/src/com/earth2me/essentials/commands/Commandmute.java [hunk 12]
```diff
@@ -50,11 +50,24 @@ public void run(final Server server, final CommandSource sender, final String co
...
                 user.setMuted(true);
+
             } else {
                 user.setMuted(!user.getMuted());
+                if (!user.getMuted ()) {
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandmute.java]: Format.

File: Essentials/src/com/earth2me/essentials/commands/Commandmute.java [hunk 13]
```diff
@@ -50,11 +50,24 @@ public void run(final Server server, final CommandSource sender, final String co
...
+                    user.setMuteReason ("");
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandmute.java]: Format.

File: Essentials/src/com/earth2me/essentials/commands/Commandmute.java [hunk 14]
```diff
@@ -66,17 +79,40 @@ public void run(final Server server, final CommandSource sender, final String co
 
             if (muted) {
                 if (muteTimestamp > 0) {
-                    sender.sendMessage(tl("mutedPlayerFor", user.getDisplayName(), muteTime));
-                    user.sendMessage(tl("playerMutedFor", muteTime));
+                    if (user.getMuteReason ().equals ("")) {
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandmute.java]: Format.

Comment (mdcfe) [reply Essentials/src/com/earth2me/essentials/commands/Commandmute.java]: Also, is there any reason for not using `hasMuteReason`?

File: Essentials/src/com/earth2me/essentials/commands/Commandmute.java [hunk 15]
```diff
@@ -66,17 +79,40 @@ public void run(final Server server, final CommandSource sender, final String co
...
+                        sender.sendMessage(tl("mutedPlayerFor", user.getDisplayName(), muteTime));
+                        user.sendMessage(tl("playerMutedFor", muteTime));
+                    } else {
+                        sender.sendMessage(tl("mutedPlayerForReason", user.getDisplayName(), muteTime, user.getMuteReason()));
+                        user.sendMessage(tl("playerMutedForReason", muteTime, user.getMuteReason()));
+                    }
                 } else {
-                    sender.sendMessage(tl("mutedPlayer", user.getDisplayName()));
-                    user.sendMessage(tl("playerMuted"));
+                    if (user.getMuteReason ().equals ("")) {
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandmute.java]: Format and `hasMuteReason` again.

File: Essentials/src/com/earth2me/essentials/commands/Commandmute.java [hunk 16]
```diff
@@ -66,17 +79,40 @@ public void run(final Server server, final CommandSource sender, final String co
...
+                        sender.sendMessage(tl("mutedPlayer", user.getDisplayName()));
+                        /** Send the player a message, why they were muted **/
+                        user.sendMessage(tl("playerMuted"));
+                    }
+                    else {
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandmute.java]: Format.

File: Essentials/src/com/earth2me/essentials/commands/Commandmute.java [hunk 17]
```diff
@@ -66,17 +79,40 @@ public void run(final Server server, final CommandSource sender, final String co
...
+                        sender.sendMessage(tl("mutedPlayerReason", user.getDisplayName(), user.getMuteReason()));
+                        /** Send the player a message, why they were muted **/
+                        user.sendMessage(tl("playerMutedReason", user.getMuteReason()));
+                    }
                 }
                 final String message;
                 if (muteTimestamp > 0) {
-                    message = tl("muteNotifyFor", sender.getSender().getName(), user.getName(), muteTime);
+                    if (user.getMuteReason ().equals ("")) {
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandmute.java]: `hasMuteReason` and format.

File: Essentials/src/com/earth2me/essentials/commands/Commandmute.java [hunk 18]
```diff
@@ -66,17 +79,40 @@ public void run(final Server server, final CommandSource sender, final String co
...
+                        message = tl("muteNotifyFor", sender.getSender().getName(), user.getName(), muteTime);
+                    }
+                    else {
+                        message = (tl("muteNotifyForReason", sender.getSender().getName(), user.getName(), muteTime, user.getMuteReason()));
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandmute.java]: Why is the `tl` call wrapped in brackets?

File: Essentials/src/com/earth2me/essentials/commands/Commandmute.java [hunk 19]
```diff
@@ -66,17 +79,40 @@ public void run(final Server server, final CommandSource sender, final String co
...
+                        user.sendMessage(tl("playerMutedReason", user.getMuteReason()));
+                    }
                 }
                 final String message;
                 if (muteTimestamp > 0) {
-                    message = tl("muteNotifyFor", sender.getSender().getName(), user.getName(), muteTime);
+                    if (user.getMuteReason ().equals ("")) {
+                        message = tl("muteNotifyFor", sender.getSender().getName(), user.getName(), muteTime);
+                    }
+                    else {
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandmute.java]: Format.

File: Essentials/src/com/earth2me/essentials/commands/Commandmute.java [hunk 20]
```diff
@@ -66,17 +79,40 @@ public void run(final Server server, final CommandSource sender, final String co
...
+                        message = (tl("muteNotifyForReason", sender.getSender().getName(), user.getName(), muteTime, user.getMuteReason()));
+                    }
                 } else {
-                    message = tl("muteNotify", sender.getSender().getName(), user.getName());
+                    if (user.getMuteReason ().equals ("")) {
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandmute.java]: Format.

File: Essentials/src/com/earth2me/essentials/commands/Commandmute.java [hunk 21]
```diff
@@ -66,17 +79,40 @@ public void run(final Server server, final CommandSource sender, final String co
...
+                        message = tl("muteNotify", sender.getSender().getName(), user.getName());
+                    }
+                    else {
+                        message = (tl("muteNotifyReason", sender.getSender().getName(), user.getName(), user.getMuteReason()));
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandmute.java]: Brackets around `tl` again.

File: Essentials/src/com/earth2me/essentials/commands/Commandwhois.java [hunk 22]
```diff
@@ -56,8 +56,9 @@ public void run(final Server server, final CommandSource sender, final String co
             sender.sendMessage(tl("whoisAFK", tl("false")));
         }
         sender.sendMessage(tl("whoisJail", (user.isJailed() ? user.getJailTimeout() > 0 ? DateUtil.formatDateDiff(user.getJailTimeout()) : tl("true") : tl("false"))));
-        sender.sendMessage(tl("whoisMuted", (user.isMuted() ? user.getMuteTimeout() > 0 ? DateUtil.formatDateDiff(user.getMuteTimeout()) : tl("true") : tl("false"))));
-
+        sender.sendMessage(user.hasMuteReason() ? tl("whoisMutedReason", (user.isMuted() ? user.getMuteTimeout() > 0 ?
+                DateUtil.formatDateDiff(user.getMuteTimeout()) : tl("true") : tl("false")),user.getMuteReason()) : tl("whoisMuted", (user.isMuted() ?
+                user.getMuteTimeout() > 0 ? DateUtil.formatDateDiff(user.getMuteTimeout()) : tl("true") : tl("false"))));
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandwhois.java]: There is an actual reason to expand these ternary statements into `if ... else`s here, as they are quite complex.

**Review by mdcfe — CHANGES_REQUESTED** [review]

**Comment by drtshock** [comment]
You should also expose this mute reason in /seen https://github.com/EssentialsX/Essentials/blob/2.x/Essentials/src/com/earth2me/essentials/commands/Commandseen.java#L120

File: Essentials/src/com/earth2me/essentials/UserData.java (hunk 9 above)
Comment (LCookman) [reply Essentials/src/com/earth2me/essentials/UserData.java]: This is a good point, Initially we had it here since we didn't have `hasMuteReason` but since we now have the ability to check, this is obsolete. I will make the change.

**Comment by drtshock** [comment]
Also a note, you could join discord to chat about these changes if you'd like. Link is in the readme.

**Comment by LCookman** [comment]
Alright, so I fixed (to my knowledge) all of the formatting issues dealing with the spacing issues, if there are more let me know so I can be sad that I missed them. I have also Updated the instances of checking the muteReason within Commandmute to use `hasMuteReason` which was also updated to check for null instead of an empty string. Finally I added the ability to check if a player is muted while offline by doing /seen while the player is Offline. This is just the same bit of code that is within the seenOnline method that checks if the player has a muteReason. If there is anything else let us know we are happy to keep refining these changes!

**Comment by LCookman** [comment]
Just curious, but would you like us to fix those Merge Conflicts with the property files?

**Comment by mdcfe** [comment]
@CreedTheFreak If possible, yes. I'd suggest rebasing onto master, pasting the added messages at the bottom then using the bash command at the top of #1734 to sort the messages.

**Comment by mdcfe** [comment]
Whoops, didn't mean to close this.

**Comment by LCookman** [comment]
Alright thanks! We will get to work on this, when we have some free time.

**Comment by LCookman** [comment]
Okay, so we were finally able to resolve those conflicts with the rebase and then sorted the files with the given command. Let us know if there is anything else that we need to do. Looks like our commits skyrocketed also, not sure if that was a byproduct of the rebase or not.

**Review by mdcfe — APPROVED** [review]

File: Essentials/src/messages.properties [hunk 23]
```diff
@@ -20,10 +20,10 @@ antiBuildPlace=\u00a74You are not permitted to place\u00a7c {0} \u00a74here.
 antiBuildUse=\u00a74You are not permitted to use\u00a7c {0}\u00a74.
 autoAfkKickReason=You have been kicked for idling more than {0} minutes.
 backAfterDeath=\u00a76Use the /back command to return to your death point.
+backUsageMsg=\u00a76Returning to previous location.
```
Comment (drtshock) [inline Essentials/src/messages.properties]: You guys are changing a lot of messages here that don't seem relevant to the PR. Can you guys try to minimize the diff in the messages files?

**Review by drtshock — CHANGES_REQUESTED** [review]

**Comment by drtshock** [comment]
It looks like the messages.properties files all have huge diffs on keys that aren't relevant to this PR. If you guys can fix up the diff for those, then I approve :D

**Comment by mdcfe** [comment]
Those diffs might be related to #1734 rather than intentional changes here.

**Comment by drtshock** [comment]
Probably, but the PR should be updated to reflect that.

File: Essentials/src/com/earth2me/essentials/commands/Commandmute.java [hunk 24]
```diff
@@ -50,11 +50,23 @@ public void run(final Server server, final CommandSource sender, final String co
...
             if (args.length > 1) {
-                final String time = getFinalArg(args, 1);
-                muteTimestamp = DateUtil.parseDateDiff(time, true);
+                final String time = args[1];
+                String muteReason;
+                try {
+                    muteTimestamp = DateUtil.parseDateDiff(time, true);
+                    muteReason = getFinalArg(args, 2);
+                } catch (Exception e) {
+                    user.setMuted((!user.getMuted()));
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandmute.java]: What's the purpose of this line?

File: Essentials/src/messages.properties [hunk 25]
```diff
@@ -46,8 +46,8 @@ broadcast=\u00a7r\u00a76[\u00a74Broadcast\u00a76]\u00a7a {0}
 buildAlert=\u00a74You are not permitted to build.
 bukkitFormatChanged=Bukkit version format changed. Version not checked.
 burnMsg=\u00a76You set\u00a7c {0} \u00a76on fire for\u00a7c {1} seconds\u00a76.
-cannotStackMob=\u00a74You do not have permission to stack multiple mobs.
```
Comment (mdcfe) [inline Essentials/src/messages.properties]: Your script seems to have different opinions on how to alphabetically sort the keys - if you could minimise the changes to only include the new keys then that would be ideal (perhaps a revert and manually add new keys in the right places).

**Review by mdcfe — CHANGES_REQUESTED** [review]
Sorry if it seems like we're messing you around a bit, but just a couple of last minute things I noticed. Once that's sorted and the others approve, we can merge.

File: Essentials/src/com/earth2me/essentials/commands/Commandmute.java (hunk 24 above)
Comment (LCookman) [reply Essentials/src/com/earth2me/essentials/commands/Commandmute.java]: I don't think there is one now that I think about it. Since regardless of if an exception is caught, we are setting the user to muted after the try catch block. We will remove this and then I believe we also have a fix for the weird property diffs. We ran the command that was given and apparently it sorted them in a different order. If this fix doesn't work we will probably have to do it the long way.

**Comment by delbertina** [comment]
I believe the weird diffs and the redundant line are fixed now. The issue with the sort may have been some OS specific stuff, but it looks all good now to me. I am seeing some weird diffs at the end of the cs, fr, it, zh, and pl message files. Tried a few things to remedy that, but no dice. If that issue is something we need to fix, may be able to dig into things deeper and find a solution. :)

**Review by drtshock — APPROVED** [review]

**Review by mdcfe — APPROVED** [review]
Everything looks good to me now, so I'd say we can now squash and merge this. Thanks for the contribution! (and for putting up with our nitpicking :p)

**Comment by LCookman** [comment]
Haha no worries, it was kind of our first time as a group working on an open source project. It was fun and we hope to contribute more in the future, without all the simple mistakes :P

**Comment by mdcfe** [comment]
@LCookman @delbertina Thanks for the contribution (and apologies for the long delay)! This feature is now available in the [builds 711 and above](https://ci.ender.zone/job/EssentialsX).
